File: apps/api/src/services/struct_fields.py
"""The declared fields of a struct property (Foundry ``object-link-types``
p.149-150; db 0064).

> "A struct is an Ontology property base type that allows users to create
> schema-based properties with multiple fields." (p.149)

**The schema is the whole difference between a struct and a `json`.** This
platform has accepted a composite value in a property since 0003 - `json` is
the escape hatch, "deliberately unconstrained" in `property_values`' own words -
so a struct is not a new *value*, it is a new *claim about* a value. Everything
here exists to make that claim checkable: what fields there are, what each one
holds, and what a value must look like to satisfy it.

**This module validates a declaration; it does not coerce a value.** Coercing
lives in ``property_values``, because the sync path and the action path both
write struct values and that module is the one mirrored into the worker. The
split is the same one ``value_format`` makes for the same reason: a rule
enforced on one write path and not the other is not a rule.

p.149 states four constraints, and all four are refusals rather than
conventions:

* **depth of one** - "Structs have a depth of one and cannot be nested", so a
  field may not itself be a struct;
* **at least one field** - a struct with none is a `json` property with a more
  specific name;
* a fixed list of **field types**, narrower than this platform's property
  types (see ``FIELD_TYPES``);
* fields are ordered and named, and a name is unique within its struct.
"""
from __future__ import annotations

import re
from typing import Any

#: The same shape a property api_name has (`0003`), and deliberately the same
#: rule rather than a looser one: a field name is read the way a property name
#: is - typed into a transform, pasted out of an error - and two naming
#: conventions one level apart is a thing to get wrong rather than a freedom.
_FIELD_API_RE = re.compile(r"^[a-z][a-z0-9_]{0,99}$")

#: p.149's list, in this platform's vocabulary. Foundry names twelve SQL types
#: (BOOLEAN, BYTE, DATE, DECIMAL, DOUBLE, FLOAT, GEOPOINT, INTEGER, LONG,
#: SHORT, STRING, TIMESTAMP); this platform's property types collapse the four
#: integer widths to `integer` and the three float widths to `float`, which is
#: the same collapse `_DUCK_TO_PROPERTY` already makes when reading a dataset.
#:
#: **What is absent is absent for a reason, one reason each**, because a reader
#: comparing this list with ``PROPERTY_TYPES`` will otherwise read the gaps as
#: an oversight:
#:
#: * `struct` - p.149's own sentence. Nesting is refused, and refused with its
#:   own message rather than as an unknown type, because somebody who has read
#:   p.58's address example will try it.
#: * `json` - it is the untyped escape hatch, so a field holding one would be
#:   nesting with the type system switched off: the value could contain
#:   anything, including another struct, and nothing would say so.
#: * `attachment` - an attachment is a reference to a stored blob, and the
#:   upload path (`POST /attachments`, §39) hands back a reference for *a
#:   property*. Making one addressable inside a struct means a second answer to
#:   "which property does this file belong to", which is a storage question and
#:   not a schema one.
#: * `time_series` - its points live where `object_type_series` says, and that
#:   row names a `property_api_name`. A series inside a struct would need that
#:   table to grow a second identity for its target, which is the thing 0047's
#:   own comment declines to introduce.
FIELD_TYPES = (
    "boolean", "date", "float", "geopoint", "integer", "string", "timestamp",
)

_FIELD_KEYS = ("api_name", "display_name", "description", "data_type")


class StructFieldError(ValueError):
    """A struct declaration that could not be stored."""


def parse(value: Any, *, data_type: str, property_name: str) -> list[dict[str, Any]] | None:
    """Normalise and check one property's ``struct_fields``.

    Returns the stored form - an ordered list of
    ``{api_name, display_name, description, data_type}`` - or ``None`` for a
    property that is not a struct. Normalised rather than passed through for
    ``value_format.parse``'s reason: what is stored has to be what was checked,
    or the two are only accidentally the same.

    **A non-struct property carrying fields is refused, not ignored.** A schema
    on a `string` property is a claim nothing reads, and silently dropping it
    would leave somebody looking at a saved definition that does not contain
    what they typed.
    """
    if data_type != "struct":
        if _stated(value):
            raise StructFieldError(
                f"{property_name!r} is a {data_type} and cannot have struct fields"
            )
        return None
    if not isinstance(value, list) or not value:
        # p.149: "Structs must have at least 1 field."
        raise StructFieldError(
            f"struct property {property_name!r} needs at least one field"
        )

    out: list[dict[str, Any]] = []
    seen: set[str] = set()
    for index, raw in enumerate(value):
        if not isinstance(raw, dict):
            raise StructFieldError(
                f"field {index + 1} of {property_name!r} is not an object"
            )
        unknown = sorted(set(raw) - set(_FIELD_KEYS))
        if unknown:
            raise StructFieldError(
                f"field {index + 1} of {property_name!r} has unknown keys: "
                + ", ".join(unknown)
            )
        api = str(raw.get("api_name") or "")
        if not _FIELD_API_RE.match(api):
            raise StructFieldError(
                f"invalid struct field name {api!r} on {property_name!r}"
            )
        if api in seen:
            raise StructFieldError(
                f"duplicate struct field {api!r} on {property_name!r}"
            )
        seen.add(api)
        field_type = str(raw.get("data_type") or "")
        if field_type == "struct":
            # **Its own message, not "unknown type".** p.149 says structs
            # "have a depth of one and cannot be nested", and this platform
            # refusing it as a typo would read as the spec being wrong rather
            # than as the refusal it is.
            raise StructFieldError(
                f"struct field {api!r} on {property_name!r} cannot itself be a "
                "struct: structs have a depth of one and cannot be nested "
                "(object-link-types p.149)"
            )
        if field_type not in FIELD_TYPES:
            raise StructFieldError(
                f"struct field {api!r} on {property_name!r} has type "
                f"{field_type!r}; expected one of " + ", ".join(FIELD_TYPES)
            )
        out.append(
            {
                "api_name": api,
                "display_name": str(raw.get("display_name") or api),
                "description": str(raw.get("description") or ""),
                "data_type": field_type,
            }
        )
    return out


def _stated(value: Any) -> bool:
    """Whether the caller said anything at all.

    ``None`` and ``[]`` both mean "no fields", and neither is a mistake worth a
    refusal on a property that could not have had any.
    """
    return value is not None and value != []


def types_by_field(fields: Any) -> dict[str, str]:
    """``{field name: data type}`` for a stored declaration.

    The form every reader actually wants - coercion, rendering, the Workshop
    extract transform - built here rather than in each of them, so a stored
    declaration that has been hand-edited into a shape nobody planned for
    produces an empty mapping in one place instead of a different exception in
    three.
    """
    if not isinstance(fields, list):
        return {}
    out: dict[str, str] = {}
    for field in fields:
        if not isinstance(field, dict):
            continue
        api = field.get("api_name")
        data_type = field.get("data_type")
        if isinstance(api, str) and isinstance(data_type, str):
            out[api] = data_type
    return out

File: apps/web/src/lib/struct-fields.ts
/**
 * The rules a struct declaration has to satisfy before it is worth sending
 * (Foundry `object-link-types` p.149, p.152–158; db 0064).
 *
 * **The server owns every refusal here** — `services/struct_fields.py` is what
 * decides whether a declaration may be stored, and this file does not get a
 * vote. What it does is answer them *early*, where the answer can still be
 * changed: a dialog whose Apply button turns a filled-in form into a 422 with
 * the dialog already closed is a dialog that wastes the work. That is the same
 * split `value-format-editor`'s `incomplete` makes, and the same sentence
 * `STATUS.md` keeps writing about panels — the server owns what is *legal*,
 * the panel owns what to *offer*.
 *
 * Pure, and in `lib/` rather than beside the component, because vitest cannot
 * parse `.tsx`: a rule that lives in a component is a rule with no unit test.
 */

import type { PropertyDataType, StructField } from "@/lib/types";

/**
 * p.149's field types, in this platform's vocabulary.
 *
 * **A mirror of `struct_fields.FIELD_TYPES` on the server, and mirrors go
 * stale** — §191's whole lesson — so `test_struct_properties.py` scans this
 * file and compares the two. The check is against the *server's* list rather
 * than a second copy of this one, which is the direction that catches an
 * addition rather than only a disagreement.
 */
export const FIELD_TYPES: PropertyDataType[] = [
  "boolean", "date", "float", "geopoint", "integer", "string", "timestamp",
];

/** A field name follows the property naming rule, one level down (db 0003). */
const FIELD_NAME = /^[a-z][a-z0-9_]{0,99}$/;

/**
 * A field name, from whatever was typed — the same normalisation the property
 * row above it applies to a property name, one level down.
 *
 * Deliberately **not** the whole rule. It lower-cases, collapses punctuation
 * and joins on underscores, which covers "Postal Code" and "postal-code"; it
 * cannot fix a name that starts with a digit, because there is nothing to
 * turn `1st_line` into that the author would recognise as what they typed. So
 * `problem` still checks, and that branch is reachable through this box rather
 * than only through the API — which is the difference between a second guard
 * and a dead one (§213).
 */
export function toFieldApiName(typed: string): string {
  const words = typed.match(/[A-Za-z0-9]+/g) ?? [];
  return words.map((w) => w.toLowerCase()).join("_").slice(0, 100);
}

/** A new row, typed `string` because that is what most fields are. */
export function blankField(): StructField {
  return { api_name: "", display_name: "", description: "", data_type: "string" };
}

/**
 * Why this declaration cannot be saved yet, or `null`.
 *
 * One sentence at a time, and the first one only: a list of every complaint
 * about a half-typed form is a list somebody has to read to find the one they
 * have already fixed.
 */
export function problem(fields: StructField[]): string | null {
  // p.149: "Structs must have at least 1 field." A struct with none is a
  // `json` property with a more specific name.
  if (fields.length === 0) return "A struct needs at least one field.";
  const seen = new Set<string>();
  for (const [index, field] of fields.entries()) {
    const name = field.api_name.trim();
    if (!name) return `Field ${index + 1} needs a name.`;
    if (!FIELD_NAME.test(name)) {
      return `${name} is not a valid field name: lower case, digits and underscores, starting with a letter.`;
    }
    if (seen.has(name)) return `Two fields are both named ${name}.`;
    seen.add(name);
  }
  return null;
}

/**
 * Which fields have been renamed, comparing what is open against what was
 * loaded — p.158's warning, in this platform's terms.
 *
 * Foundry's version is about a RID: *"changing a struct field's API name will
 * result in a new struct field RID being generated … Any applications that
 * reference the updated struct field will need to be updated as well."* This
 * platform has no RID, and the consequence lands somewhere else instead: a
 * stored value is a mapping keyed by field name, so after a rename every
 * instance still holds the old key and the field reads as empty **until the
 * next sync**. Same warning, different mechanism, and the mechanism is what
 * makes it actionable — the fix is a sync, and saying so is the point.
 *
 * Matched **by position**, because that is the only identity a row in this
 * dialog has: a field is not addressable by the name being edited, and the
 * list is short and hand-ordered. A field added or removed shifts the
 * comparison, which is why an added row (no previous name) is not a rename.
 */
export function renamedFields(
  before: StructField[] | null | undefined,
  after: StructField[],
): string[] {
  const previous = before ?? [];
  const out: string[] = [];
  for (const [index, field] of after.entries()) {
    const was = previous[index]?.api_name;
    const now = field.api_name.trim();
    if (was && now && was !== now) out.push(was);
  }
  return out;
}

/**
 * What to show for a struct value when the declaration is to hand.
 *
 * `[[display name, rendered value], …]` in the **declared** order, which is
 * the order the value is stored in (`_coerce_struct` rebuilds it) and the one
 * p.154 says the author chose. A field the value has nothing for still gets a
 * row: p.59's "semantic grouping" is the point of the type, and an address
 * that silently omits its postcode looks like an address without one rather
 * than like a value that is missing it.
 *
 * Returns `null` for anything that is not a struct value against these fields,
 * so a caller can fall back to whatever it did before rather than render an
 * empty list.
 */
export function structRows(
  fields: StructField[] | null | undefined,
  value: unknown,
): [string, unknown][] | null {
  if (!fields || fields.length === 0) return null;
  if (typeof value !== "object" || value === null || Array.isArray(value)) return null;
  const held = value as Record<string, unknown>;
  return fields.map((field) => [
    field.display_name || field.api_name,
    held[field.api_name] ?? null,
  ]);
}
